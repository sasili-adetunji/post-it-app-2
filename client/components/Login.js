import React from 'react';
import mui from 'material-ui';
import { Link } from 'react-router-dom';
import GoogleButton from 'react-google-button';
import { Card, CardTitle } from 'material-ui/Card';
import TextField from 'material-ui/TextField';
import RaisedButton from 'material-ui/RaisedButton';
import FlatButton from 'material-ui/FlatButton';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import PostItActions from '../actions/PostItActions';
import PostItStore from '../stores/PostItStore';

/**
 * creates Login componets
 * @class Login
 * @extends {React.Component}
 */
class Login extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      errors: ''
    };
    this.onChange = this.onChange.bind(this);
    this.onClick = this.onClick.bind(this);
    this.onClickGoogle = this.onClickGoogle.bind(this);
    this.onClickReset = this.onClickReset.bind(this);
  }
  /**
     * Monitors changes in the components and change the state
     * @param {object} e
     * @returns {void}
     * @memberOf Login
  */
  onChange(e) {
    this.setState({
      [e.target.name]: e.target.value
    });
  }

  /**
     * Makes an action call to log a user with email and password
     * @param {object} e
     * @returns {void}
     * @memberOf Login
  */
  onClick(event) {
    event.preventDefault();
    const user = {
      email: this.state.email,
      password: this.state.password
    };
    if ((!user.email) || (!user.password)) {
      this.setState({
        errors: 'Please enter a valid email and password'
      });
    } else {
      PostItActions.login(user);
      this.setState({
        email: '',
        password: '',
        errors: ''
      });
    }
  }
  /**
     * Makes an action call to log a user with google
     * @param {object} e
     * @returns {void}
     * @memberOf Login
  */
  onClickGoogle() {
    PostItActions.googleLogin();
  }
  /**
     * Makes an action call to reset password
     * @param {object} e
     * @returns {void}
     * @memberOf Login
  */
  onClickReset(event) {
    event.preventDefault();
    const email =
      {
        email: this.state.email
      };
    PostItActions.resetPassword(email);
  }

  /**
     * @returns {String} The HTML markup for the Login
     * @memberOf Login
     */
  render() {
    return (
      <div>
        <MuiThemeProvider >
          <Card className="card" >
            <CardTitle
              title="Login Form"
              subtitle="To continue using PostIt, you need to login below" />
            <TextField
              name="email" onChange={this.onChange} value={this.state.email}
              floatingLabelText="Your Email" /><br />
            <TextField
              name="password" onChange={this.onChange} value={this.state.password}
              floatingLabelText="Your Password"
              type="password" /><br />
            <br />
            <span className="error"> {this.state.errors} {PostItStore.getErrors()} </span> <br />
            <p> Already Have an account,<Link to="/signup"> Register here </Link> </p>

            <p> Forgot your Password? Enter your Email and <a
              href="/#/signup"
              onClick={this.onClickReset}> Click here </a> </p>
            <RaisedButton
              label="Login" primary onClick={this.onClick} />
            <br />
            <br />
            <div className="row">
              <center>
                <GoogleButton onClick={this.onClickGoogle} />
              </center>
            </div>
          </Card>
        </MuiThemeProvider>
      </div>
    );
  }
}
export default Login;
