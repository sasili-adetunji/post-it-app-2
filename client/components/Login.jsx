import React from 'react';
import firebase from 'firebase';
import { Link } from 'react-router-dom';
import GoogleButton from 'react-google-button';
import PostItActions from '../actions/PostItActions';
import PostItStore from '../stores/PostItStore';
import config from '../../server/app/config/database';
import NavBar from './NavBar';

/**
 * 
 * @description gets user data and login a user
 * @export
 * @param {object} props
 * @class Login
 * @extends {Component}
 */

class Login extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      errors: {},
      isLoading: false,
    };
    this.onChange = this.onChange.bind(this);
    this.onClick = this.onClick.bind(this);
    this.onClickGoogle = this.onClickGoogle.bind(this);
  }
   /**
    * @method onChange
    *
    * @description Monitors changes in the components and change the state
    *
    * @memberof Login
    *
    * @param {object} event
    *
    * @returns {void}
    */
  onChange(event) {
    this.setState({
      [event.target.name]: event.target.value
    });
  }

/**
 * @description Makes an action call to Sign in a user with email and password
 * 
 * @param {object} event
 * 
 * @returns {void}
 * 
 * @memberof Login
*/
  onClick(event) {
    event.preventDefault();
    const user = {
      email: this.state.email,
      password: this.state.password
    };
    if (!user.email) {
      this.setState({ errors: { email: 'Email is required' } })
    } else if (!user.password) {
        this. setState({ errors: { password: 'Password is required' } })
    } else {
    PostItActions.login(user);
    this.setState({
      email: '',
      password: '',
      errors: '',
      success: '',
      isLoading: true
    });
  }
  }
   /**
     * @description Makes an action call to Sign in a user with with google
     * @param {object} event
     * @returns {void}
     * @memberof Lofin
  */
  onClickGoogle(event) {
    event.preventDefault();
    firebase.initializeApp(config);
    const provider = new firebase.auth.GoogleAuthProvider();
    provider.addScope('https://www.googleapis.com/auth/plus.login');
    firebase.auth().signInWithPopup(provider)
    .then((result) => {
      PostItActions.googleLogin(result);
      this.setState({
        isLoading: true
      });
    })
  }
 

   /**
   * @method render
   * Render react component
   * 
   * @returns {String} The HTML markup for the Register
   * @memberof Register
   */

  render() {
    const isLoading = () => {
      const loading = (
        this.state.isLoading ? <div id="loader"></div> : <span></span>
      );
      return loading;
    }
    return (
      <div>
        <NavBar />
        <div className="login-container">
            {isLoading()}
          <h1>Login</h1>
          <p>To continue using PostIt, you need to login below</p>
        <div className='error'> {this.state.errors.email}  
          {this.state.errors.password}   
      </div>
          <form>
            <div className="form-group">
              <label>Email address</label>
              <input type="email" className="form-control" 
              name="email" onChange={this.onChange} 
              value={this.state.email} placeholder="Email" />
            </div>
            <div className="form-group">
              <label>Password</label>
              <input type="password" className="form-control" 
               name="password" onChange={this.onChange} 
              value={this.state.password} placeholder="Password" />
            </div>
            <button type="submit" className="btn btn-default"
            onClick={this.onClick} >Submit</button>
          </form>
          <div className="clear"> </div>
           <p> <Link to="/forgotPassword"> Forgot Password? 
            </Link> </p>
          <div className="clear"> </div>
          <GoogleButton onClick={this.onClickGoogle} />
          <br />
          <div className="clear"> </div>
          <p> Don't have an account? <Link to="/signup"> Register here 
            </Link> </p>
        </div>
      </div>
    );
  }
}
export default Login;
