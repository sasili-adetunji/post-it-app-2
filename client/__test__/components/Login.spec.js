import React from 'react';
import { shallow, render } from 'enzyme';
import Login from '../../components/Login.jsx';


function setup() {
  const props = {
    onClick: () => {},
    onChange: () => {},
    onClickReset: () => {},
    onClickGoogle: () => {},
  };
  return shallow(<Login {...props} />);
}
describe('Login', () => {
  const component = setup();

  it('should match snapshot test', () => {
     expect(component).toMatchSnapshot();
   });
  it('should render', () => {
    expect(component).toBeDefined();
  });
  it('Should contain two div', () => {
    expect(component.find('div').length).toEqual(2);
  });
  it('Should contain two p', () => {
    expect(component.find('p').length).toEqual(2);
  });
  it('Should contain two textfields', () => {
    expect(component.find('TextField').length).toEqual(2);
  });
  it('Should contain a Login Button', () => {
    expect(component.find('RaisedButton').props().label).toEqual('Login');
    expect(component.find('RaisedButton').length).toEqual(1);
  });
  it('Should contain a Google Button', () => {
    expect(component.find('GoogleButton').props().onClick).toExist;
    expect(component.find('GoogleButton').length).toEqual(1);
  });
  it('should call function on click of login submit button', () => {
    const preventDefault = jest.fn();
    component.find('RaisedButton').simulate('click', { preventDefault });
    expect(preventDefault).toBeCalled();
  });
});
describe('Login  Test', () => {
  it('should take props', () => {
    const component = setup();
    expect(component.props().onChange).toExist;
    expect(component.props().onClick).toExist;
    expect(component.props().onClickReset).toExist;
    expect(component.props().onClickGoogle).toExist;
  });
});
