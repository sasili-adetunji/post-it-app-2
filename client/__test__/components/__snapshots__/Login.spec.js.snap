// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match snapshot test 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <MuiThemeProvider>
      <Card
        className="card"
        expandable={false}
        expanded={null}
        initiallyExpanded={false}
      >
        <CardTitle
          subtitle="To continue using PostIt, you need to login below"
          title="Login Form"
        />
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelText="Your Email"
          fullWidth={false}
          multiLine={false}
          name="email"
          onChange={[Function]}
          rows={1}
          type="text"
          underlineShow={true}
          value=""
        />
        <br />
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          floatingLabelText="Your Password"
          fullWidth={false}
          multiLine={false}
          name="password"
          onChange={[Function]}
          rows={1}
          type="password"
          underlineShow={true}
          value=""
        />
        <br />
        <br />
        <p>
           Already Have an account,
          <Link
            replace={false}
            to="/signup"
          >
             Register here 
          </Link>
           
        </p>
        <p>
           Forgot your Password? Enter your Email and 
          <a
            href="/#/signup"
            onClick={[Function]}
          >
             Click here 
          </a>
           
        </p>
        <RaisedButton
          disabled={false}
          fullWidth={false}
          label="Login"
          labelPosition="after"
          onClick={[Function]}
          primary={true}
          secondary={false}
        />
        <br />
        <br />
        <div
          className="row"
        >
          <center>
            <GoogleButton
              disabled={false}
              label="Sign in with Google"
              onClick={[Function]}
              type="dark"
            />
          </center>
        </div>
      </Card>
    </MuiThemeProvider>
  </div>,
  "nodes": Array [
    <div>
      <MuiThemeProvider>
        <Card
          className="card"
          expandable={false}
          expanded={null}
          initiallyExpanded={false}
        >
          <CardTitle
            subtitle="To continue using PostIt, you need to login below"
            title="Login Form"
          />
          <TextField
            disabled={false}
            floatingLabelFixed={false}
            floatingLabelText="Your Email"
            fullWidth={false}
            multiLine={false}
            name="email"
            onChange={[Function]}
            rows={1}
            type="text"
            underlineShow={true}
            value=""
          />
          <br />
          <TextField
            disabled={false}
            floatingLabelFixed={false}
            floatingLabelText="Your Password"
            fullWidth={false}
            multiLine={false}
            name="password"
            onChange={[Function]}
            rows={1}
            type="password"
            underlineShow={true}
            value=""
          />
          <br />
          <br />
          <p>
             Already Have an account,
            <Link
              replace={false}
              to="/signup"
            >
               Register here 
            </Link>
             
          </p>
          <p>
             Forgot your Password? Enter your Email and 
            <a
              href="/#/signup"
              onClick={[Function]}
            >
               Click here 
            </a>
             
          </p>
          <RaisedButton
            disabled={false}
            fullWidth={false}
            label="Login"
            labelPosition="after"
            onClick={[Function]}
            primary={true}
            secondary={false}
          />
          <br />
          <br />
          <div
            className="row"
          >
            <center>
              <GoogleButton
                disabled={false}
                label="Sign in with Google"
                onClick={[Function]}
                type="dark"
              />
            </center>
          </div>
        </Card>
      </MuiThemeProvider>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_context": Object {},
    "_element": <Login
      onChange={[Function]}
      onClick={[Function]}
      onClickGoogle={[Function]}
      onClickReset={[Function]}
    />,
    "_instance": Login {
      "context": Object {},
      "onChange": [Function],
      "onClick": [Function],
      "onClickGoogle": [Function],
      "onClickReset": [Function],
      "onError": [Function],
      "props": Object {
        "onChange": [Function],
        "onClick": [Function],
        "onClickGoogle": [Function],
        "onClickReset": [Function],
      },
      "refs": Object {},
      "state": Object {
        "email": "",
        "errors": "",
        "password": "",
        "success": "",
      },
      "updater": Updater {
        "_renderer": [Circular],
      },
    },
    "_newState": null,
    "_rendered": <div>
      <MuiThemeProvider>
        <Card
          className="card"
          expandable={false}
          expanded={null}
          initiallyExpanded={false}
        >
          <CardTitle
            subtitle="To continue using PostIt, you need to login below"
            title="Login Form"
          />
          <TextField
            disabled={false}
            floatingLabelFixed={false}
            floatingLabelText="Your Email"
            fullWidth={false}
            multiLine={false}
            name="email"
            onChange={[Function]}
            rows={1}
            type="text"
            underlineShow={true}
            value=""
          />
          <br />
          <TextField
            disabled={false}
            floatingLabelFixed={false}
            floatingLabelText="Your Password"
            fullWidth={false}
            multiLine={false}
            name="password"
            onChange={[Function]}
            rows={1}
            type="password"
            underlineShow={true}
            value=""
          />
          <br />
          <br />
          <p>
             Already Have an account,
            <Link
              replace={false}
              to="/signup"
            >
               Register here 
            </Link>
             
          </p>
          <p>
             Forgot your Password? Enter your Email and 
            <a
              href="/#/signup"
              onClick={[Function]}
            >
               Click here 
            </a>
             
          </p>
          <RaisedButton
            disabled={false}
            fullWidth={false}
            label="Login"
            labelPosition="after"
            onClick={[Function]}
            primary={true}
            secondary={false}
          />
          <br />
          <br />
          <div
            className="row"
          >
            <center>
              <GoogleButton
                disabled={false}
                label="Sign in with Google"
                onClick={[Function]}
                type="dark"
              />
            </center>
          </div>
        </Card>
      </MuiThemeProvider>
    </div>,
    "_rendering": false,
    "_updater": Updater {
      "_renderer": [Circular],
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Login
    onChange={[Function]}
    onClick={[Function]}
    onClickGoogle={[Function]}
    onClickReset={[Function]}
  />,
}
`;
